### PR TITLE
forkproxy: use correct types for {g,u}ids

### LINKED ISSUE #4848: LXD 3.3 build fail on musl
I'm trying to build latest LXD 3.3 release on VoidLinux which has musl cross builds. Glibc variants is ok but musl fails with error:
```
_build-lxd-xbps/src/github.com/lxc/lxd/lxd/main_forkproxy.go:593:21: could not determine kind of name for C.__gid_t
_build-lxd-xbps/src/github.com/lxc/lxd/lxd/main_forkproxy.go:605:21: could not determine kind of name for C.__uid_t
```
This is added with this commit: https://github.com/lxc/lxd/commit/4ba2f3d894d3790dceaf11097eb496faa3953233

Full build log with error is here: https://travis-ci.org/void-linux/void-packages/jobs/409734708#L4797

### PR DESCRIPTION
**Pull request by brauner**
__{g,u}id_t are internal types

Closes #4848.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
@brauner this breaks Go 1.9 apparently

**Comment by stgraber** [comment]
Not sure why that is though... easiest way out of this will likely be a tiny cgo function in the file while we call rather than calling setuid/setgid directly.

**Comment by brauner** [comment]
> @brauner this breaks Go 1.9 apparently

I already have a solution.

File: lxd/main_forkproxy.go [hunk 1]
```diff
@@ -584,27 +595,26 @@ func (c *cmdForkproxy) Run(cmd *cobra.Command, args []string) error {
...
-		errno := C.setgid(C.__gid_t(gid))
-		if errno < 0 {
-			return fmt.Errorf("setgid: %v", errno)
-		}
 	}
 
+	uid := int64(0)
 	if args[10] != "" {
-		uid, err := strconv.ParseInt(args[10], 10, 32)
+		uid, err = strconv.ParseInt(args[10], 10, 32)
```
Comment (stgraber) [inline lxd/main_forkproxy.go]: ParseUint(args[10, 10, 32)

File: lxd/main_forkproxy.go [hunk 2]
```diff
@@ -584,27 +595,26 @@ func (c *cmdForkproxy) Run(cmd *cobra.Command, args []string) error {
 	}
 
 	// Drop privilege if requested
+	gid := int64(0)
 	if args[9] != "" {
-		gid, err := strconv.ParseInt(args[9], 10, 32)
+		gid, err = strconv.ParseInt(args[9], 10, 32)
```
Comment (stgraber) [inline lxd/main_forkproxy.go]: ParseUint(args[10, 10, 32)

File: lxd/main_forkproxy.go [hunk 3]
```diff
@@ -584,27 +595,26 @@ func (c *cmdForkproxy) Run(cmd *cobra.Command, args []string) error {
 	}
 
 	// Drop privilege if requested
+	gid := int64(0)
```
Comment (stgraber) [inline lxd/main_forkproxy.go]: uint64

File: lxd/main_forkproxy.go [hunk 4]
```diff
@@ -584,27 +595,26 @@ func (c *cmdForkproxy) Run(cmd *cobra.Command, args []string) error {
...
 			return err
 		}
-
-		errno := C.setgid(C.__gid_t(gid))
-		if errno < 0 {
-			return fmt.Errorf("setgid: %v", errno)
-		}
 	}
 
+	uid := int64(0)
```
Comment (stgraber) [inline lxd/main_forkproxy.go]: uint64

File: lxd/main_forkproxy.go [hunk 5]
```diff
@@ -584,27 +595,26 @@ func (c *cmdForkproxy) Run(cmd *cobra.Command, args []string) error {
...
 		if err != nil {
 			return err
 		}
+	}
 
-		errno := C.setuid(C.__uid_t(uid))
-		if errno < 0 {
-			return fmt.Errorf("setuid: %v", errno)
+	if uid != 0 || gid != 0 {
+		ret := C.switch_uid_gid(C.int(uid), C.int(gid))
```
Comment (stgraber) [inline lxd/main_forkproxy.go]: C.uint32(uid), C.uint32(gid)

File: lxd/main_forkproxy.go [hunk 6]
```diff
@@ -43,6 +43,17 @@ int whoami = -ESRCH;
 #define FORKPROXY_PARENT 0
 #define FORKPROXY_UDS_SOCK_FD_NUM 200
 
+int switch_uid_gid(int uid, int gid)
```
Comment (stgraber) [inline lxd/main_forkproxy.go]: uint32 uid, uint32 gid

**Pull request merged**